components/blitz/src/omero/gateway/facility/ROIFacility.java

```diff
@@ -834,71 +846,193 @@ else if (shape.isDirty()) {
+                            + "left outer join fetch roi.folderLinks as links ";
+                } else if (imageId > -1) {
+                    // get rois for the image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.image as image "
+                            + "left outer join fetch roi.folderLinks as links "
+                            + "where image.id = :id";
+                    p.addId(imageId);
+
+                } else {
+                    // get rois not attached to any image
+                    query = "select distinct roi from Roi roi "
+                            + "left outer join fetch roi.folderLinks as links "
+                            + "where roi.image is null";
+                }
 
-            sb.append("select roilink from FolderRoiLink as roilink ");
-            sb.append("left outer join fetch roilink.parent as folder ");
-            sb.append("left outer join fetch roilink.child as roi ");
-            sb.append("where folder.id = :folderId ");
+                IQueryPrx service = gateway.getQueryService(ctx);
 
-            List<IObject> links = qs.findAllByQuery(sb.toString(), param);
+                List<IObject> objs = service.findAllByQuery(query, p);
 
-            Set<Long> roiIds = new HashSet<Long>();
-            for (IObject l : links) {
-                FolderRoiLink link = (FolderRoiLink) l;
-                roiIds.add(link.getChild().getId().getValue());
+                for (IObject obj : objs) {
+                    Roi roi = (Roi) obj;
+                    if (roi.copyFolderLinks().isEmpty())
+                        roiIDs.add(roi.getId().getValue());
+                }
+            } else {
+                // get all rois which are part of the given folders
+                Map<String, Collection<Long>> target = new HashMap<String, Collection<Long>>();
+                target.put("Folder", folderIds);
+
+                FindChildren finder = Requests.findChildren().target(target)
+                        .childType("Roi").build();
+
+                Map<String, String> callContext = new HashMap<String, String>();
+                callContext.put("omero.group", "" + ctx.getGroupID());
+
+                CmdCallbackI cb = gateway.submit(ctx, finder);
+                cb.block(10000);
+                FoundChildren found = (FoundChildren) cb.getResponse();
+                roiIDs = found.children.remove(ome.model.roi.Roi.class
+                        .getName());
+
+                if (imageId != null) {
+                    if (imageId > -1) {
+                        // only take ROIs for the given image into account
+                        target = new HashMap<String, Collection<Long>>();
+                        target.put("Image", Collections.singleton(imageId));
+                        finder = Requests.findChildren().target(target)
+                                .childType("Roi").build();
```

Comment:
I forgot about the masks... they are somehow special, and have to be excluded, right? If so, is there a "neat" way how to do that?